[opaque obj] Support nested opaque objs

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172099
* #172092
* __->__ #171484
* #171483

This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

angelayi committed:
[opaque obj] Support nested opaque objs

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode. I'm not sure if `CallMethodSource` is implemented correctly :3 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode. I'm not sure if `CallMethodSource` is implemented correctly :3 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/script_object.py:
```diff
@@ -266,7 +268,30 @@ def call_method_and_return_constant(real_obj):
+                )
+                return VariableBuilder(tx, call_source)(constant_val)
+            else:
+                return SourcelessBuilder.create(tx, constant_val)
```
for my understanding: why is it preferred that we have a source? Is it that we'll get graph breaks later for things?

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/script_object.py:
```diff
@@ -289,6 +290,26 @@ def call_method(
+                for r in pytree.tree_flatten(constant_val)[0]
+            ):
+                unimplemented(
+                    gb_type="Opaque object member with method-type USE_REAL returned a reference-type opaque objects.",
```
nit: "a reference-type opaque objects"

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Support nested opaque objs"


This is adding support for reference-type opaque objects that contain reference-type opaque objects. If the inner opaque object is extracted, we should lift it as an input to the program and properly set up the bytecode via the CallMethodSource. 

Specifically this occurs in DeviceMesh's `get_group` method which returns a ProcessGroup.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
